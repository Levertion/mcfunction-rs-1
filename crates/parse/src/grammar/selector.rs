use super::*;
use crate::{
    ast::GroupType::{self, *},
    parser::{
        Parser,
        StartInfo::{self, Skip},
    },
    syntax::TokenKind::*,
    TokenSet,
};

const SELECTOR_TYPE: &[(&str, GroupType)] = &[
    ("p", SelectorModP),
    ("a", SelectorModA),
    ("r", SelectorModR),
    ("s", SelectorModS),
    ("e", SelectorModE),
];

pub fn selector(p: &mut Parser) {
    let mk = p.start(Selector, StartInfo::None);
    p.expect(At);
    if !p.expect_keyword(SELECTOR_TYPE) {
        p.bump_recover(TokenSet::empty());
    }
    if p.at(LBracket) {
        let argsmk = p.start(SelectorArgument, Skip);
        p.bump();
        if !p.at(RBracket) {
            loop {
                let argmk = p.start(SelectorArgumentEntry, Skip);
                uq_string(p);
                p.expect(Eq);
                seletor_arg_value(p);
                p.finish(argmk);
                if p.at(RBracket) {
                    break;
                }
                p.expect(Comma);
                if p.at(Eof) {
                    break;
                }
            }
        }
        p.expect(RBracket);
        p.finish(argsmk);
    }
    p.finish(mk);
}

pub fn seletor_arg_value(p: &mut Parser) {
    p.eat(Excl);
    if p.at(QuotedString) {
        p.bump();
    } else if p.at(LCurly) {
        let mapmk = p.start(SelectorArgumentMap, Skip);
        p.bump();
        if !p.at(RCurly) {
            loop {
                let argmk = p.start(SelectorArgumentMapEntry, Skip);
                resource_location(p);
                p.expect(Eq);
                seletor_arg_value(p);
                p.finish(argmk);
                if p.at(RCurly) {
                    break;
                }
                p.expect(Comma);
                if p.at(Eof) {
                    break;
                }
            }
        }
        p.expect(RCurly);
        p.finish(mapmk);
    } else {
        if !try_range(p) {
            resource_location(p);
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::{parse, testing::format_astnode};

    use insta::assert_snapshot;

    macro_rules! selector_test {
        ($name:ident, $e:expr) => {
            #[test]
            fn $name() {
                assert_snapshot!(parse_selector($e));
            }
        };
    }

    fn parse_selector(i: &str) -> String {
        format_astnode(&parse(i, super::selector), 0)
    }

    selector_test!(no_args, "@p");
    selector_test!(empty_args, "@e[]");
    selector_test!(args_word, "@s[tag=hello]");
    selector_test!(args_range, "@r[distance=1..17.5]");
    selector_test!(args_map_empty, "@a[scores={}]");
    selector_test!(
        args_map_rid,
        "@p[advancements={path/to/adv={criteria=false}}]"
    );
    selector_test!(args_map_range, "@e[scores={myobjective=-12..74}]");
    selector_test!(args_multiple, "@s[tag=hello,tag=goodbye,scores={}]");
    selector_test!(arg_invert, "@r[type=!minecraft:pig]");

    selector_test!(invalid_mod, "@q[name=\"hello\"]");
    selector_test!(unclosed_arg_empty, "@e[");
    selector_test!(unclosed_arg_key, "@p[type");
    selector_test!(arg_unclosed_map_nokey, "@s[score={");
    selector_test!(arg_unclosed_map_noeq, "@p[advancements={hello");
}
