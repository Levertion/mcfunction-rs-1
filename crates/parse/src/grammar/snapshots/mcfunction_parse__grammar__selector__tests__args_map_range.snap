---
source: crates/parse/src/grammar/selector.rs
expression: "parse_selector(\"@e[scores={myobjective=-12..74}]\")"
---
Group(Selector) at 0:0 - 0:32 {
    Token(At) `@` at 0:0 - 0:1
    Joined(SelectorModE) `e` at 0:1 - 0:2
    Group(SelectorArgument) at 0:2 - 0:32 {
        Token(LBracket) `[` at 0:2 - 0:3
        Group(SelectorArgumentEntry) at 0:3 - 0:31 {
            Joined(UnquotedString) `scores` at 0:3 - 0:9
            Token(Eq) `=` at 0:9 - 0:10
            Group(SelectorArgumentMap) at 0:10 - 0:31 {
                Token(LCurly) `{` at 0:10 - 0:11
                Group(SelectorArgumentMapEntry) at 0:11 - 0:30 {
                    Joined(ResourceLocation) `myobjective` at 0:11 - 0:22
                    Token(Eq) `=` at 0:22 - 0:23
                    Group(Range) at 0:23 - 0:30 {
                        Joined(Float) `-12` at 0:23 - 0:26
                        Token(DotDot) `..` at 0:26 - 0:28
                        Joined(Float) `74` at 0:28 - 0:30
                    }
                }
                Token(RCurly) `}` at 0:30 - 0:31
            }
        }
        Token(RBracket) `]` at 0:31 - 0:32
    }
}

